tests/MongoDB.Driver.Tests/AsyncCursorTests.cs

```diff
@@ -31,6 +33,77 @@ namespace MongoDB.Driver.Tests
     public class AsyncCursorTests
     {
         //public methods
+        [SkippableTheory]
+        [ParameterAttributeData]
+        public void Cursor_should_call_close_in_thread_safe_way(
+            [Values(false, true)] bool cappedCollection,
+            [Values(false, true)] bool async)
+        {
+            string testCollectionName = "test";
+            string testDatabaseName = "test";
+            var client = CreateClient();
+            DropCollection(client, testDatabaseName, testCollectionName);
+            var database = client.GetDatabase(testDatabaseName);
+
+            var createCollectionOptions = new CreateCollectionOptions();
+            if (cappedCollection)
+            {
+                createCollectionOptions.Capped = true;
+                createCollectionOptions.MaxSize = 1000;
+            }
+
+            database.CreateCollection(testCollectionName, createCollectionOptions);
+            var collection = database.GetCollection<BsonDocument>(testCollectionName);
+            collection.InsertOne(new BsonDocument());
+
+            var findOptions = new FindOptions()
+            {
+                BatchSize = 1
+            };
+            if (cappedCollection)
+            {
+                findOptions.CursorType = CursorType.TailableAwait;
+            }
+            var findResult = collection.Find(FilterDefinition<BsonDocument>.Empty, findOptions);
+            IAsyncCursor<BsonDocument> cursor;
+            if (async)
+            {
+                cursor = findResult
+                    .ToCursorAsync()
+                    .GetAwaiter()
+                    .GetResult();
+            }
+            else
+            {
+                cursor = findResult.ToCursor();
+            }
+
+            var asyncCursor = cursor.As<AsyncCursor<BsonDocument>>();
+            var enumerable = asyncCursor
+                .ToEnumerable()
+                .GetEnumerator();
+
+            enumerable.MoveNext();
+
+            var thread = new Thread(
+                () =>
+                {
+                    Thread.Sleep(1000);
+                    asyncCursor.Close();
+                });
+            thread.Start();
+            var exception = Record.Exception(() => enumerable.MoveNext());
+
+            if (cappedCollection)
+            {
+                exception.Should().BeOfType<ObjectDisposedException>();
+            }
+            else
+            {
+                exception.Should().BeNull();
+            }
+        }
```

Comment:
I see no reason the exception should be any different. If `Close` has been closed then `MoveNext` MUST throw an exception. Lines 95-105 should just be:

    exception.Should().BeOfType<ObjectDisposedException>();

The only reason you are seeing different exceptions is because you are not waiting for the other thread to call `Close` before calling `MoveNext`.